cmake/Modules/LibtorrentMacros.cmake

```diff
@@ -1,80 +1,210 @@
-# Various helper function and macros for building libtorrent
-
+# Various helper functions and macros for building libtorrent
 include(FeatureSummary)
 
-# macro for issuing option() and add_feature_info() in a single call.
+# function for issuing option() and add_feature_info() in a single call.
+# it is used to define boolean options. A summary of the values of these options can then be printed
+# with the native FeatureSummary module
 # Synopsis:
-# feature_option(<option_and_feature_name> <option_and_feature_description> <default_option_value>)
-macro(feature_option _name _description _default)
-	option(${_name} "${_description}" ${_default})
-	add_feature_info(${_name} ${_name} "${_description}")
-endmacro()
+# feature_option(<option/feature name> <description> <default value>)
+function(feature_option)
+	set(options "")
+	set(oneValueArgs NAME DEFAULT DESCRIPTION)
+	set(multiValueArgs "")
+
+	cmake_parse_arguments(
+		PARSE_ARGV 0
+		ARG
+		"${options}" "${oneValueArgs}" "${multiValueArgs}"
+	)
+
+	string(CONCAT _DESCRIPTION "${ARG_DESCRIPTION} (default: ${ARG_DEFAULT})")
+	option("${ARG_NAME}" "${_DESCRIPTION}" "${ARG_DEFAULT}")
+	add_feature_info("${ARG_NAME}" "${ARG_NAME}" "${_DESCRIPTION}")
+endfunction()
 
-# function to add a simple build option which controls compile definition(s) for a target.
+# function for robustly defining multi-valued cache options
 # Synopsis:
-# target_optional_compile_definitions(<target> [FEATURE]
-#   NAME <name> DESCRIPTION <description> DEFAULT <default_value>
-#   [ENABLED [enabled_compile_definitions...]]
-#   [DISABLED [disabled_compile_defnitions...]]
-# )
-# NAME, DESCRIPTION and DEFAULT are passed to option() call
-# if FEATURE is given, they are passed to add_feature_info()
-# ENABLED lists compile definitions that will be set on <target> when option is enabled,
-# DISABLED lists definitions that will be set otherwise
-function(target_optional_compile_definitions _target _scope)
-	set(options FEATURE)
-	set(oneValueArgs NAME DESCRIPTION DEFAULT)
-	set(multiValueArgs ENABLED DISABLED)
-	cmake_parse_arguments(TOCD ${options} "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
-	option(${TOCD_NAME} "${TOCD_DESCRIPTION}" ${TOCD_DEFAULT})
-	if (${${TOCD_NAME}})
-		target_compile_definitions(${_target} ${_scope} ${TOCD_ENABLED})
-	else()
-		target_compile_definitions(${_target} ${_scope} ${TOCD_DISABLED})
+# feature_option_multivalue(NAME <feature name> DESCRIPTION <description>
+#	DEFAULT <default value> ALLOWED_VALUES <list of allowed values>)
+# Returns to parent scope:
+# - Appends the new option name to a list named MULTIVALUE_FEATURE_NAMES
+# - Appends the new option default to MULTIVALUE_FEATURE_DEFAULTS
+function(feature_option_multivalue)
```

Comment:
"Multivalue" is something that bears multiple values at once, while this implements choices selection.